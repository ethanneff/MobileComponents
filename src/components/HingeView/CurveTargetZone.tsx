import React, { memo } from "react";
import { Path } from "react-native-svg";
import Theme from "../../commons/Theme";
import { Point2 } from "../../commons/Utils/Maths/geometry/Point";
import { describeBezierTargetZone } from "../../commons/Utils/SvgUtils";

interface Props {
  isInTargetZone: boolean;
  startHingePosition: Point2;
  endHingePosition: Point2;
  controlPointOuter: Point2;
  controlPointInner: Point2;
}

export default memo(function CurveTargetZone({
  isInTargetZone,
  startHingePosition,
  endHingePosition,
  controlPointOuter,
  controlPointInner
}: Props) {
  const fillColor = isInTargetZone
    ? Theme.color.success
    : Theme.color.neutral50;

  const targetZoneCurve = describeBezierTargetZone(
    startHingePosition,
    controlPointOuter,
    controlPointInner,
    endHingePosition
  );

  return (
    <>
      <Path fill={fillColor} d={targetZoneCurve} />
    </>
  );
});
