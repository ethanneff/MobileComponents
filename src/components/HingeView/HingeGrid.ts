import { createPoint2 } from "../../commons/Utils/Maths/geometry/Point";

export const HingeGridPositionPoints = {
  centerLeft: createPoint2(550, 880),
  centerRight: createPoint2(1050, 880),
  centerMiddle: createPoint2(800, 880),
  topLeft: createPoint2(550, 690),
  topMiddle: createPoint2(800, 690),
  topRight: createPoint2(1050, 690),
  bottomMiddle: createPoint2(800, 1110),
  bottomRight: createPoint2(1050, 1110),
  bottomLeft: createPoint2(550, 1110)
};

export type HingeGridPosition = keyof typeof HingeGridPositionPoints;
