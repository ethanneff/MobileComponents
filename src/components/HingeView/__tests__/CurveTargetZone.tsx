import { render } from "@testing-library/react-native";
import React from "react";
import {
  calculateControlPoints,
  getMidPointPosition
} from "../../CurveLineView/utils";
import CurveTargetZone from "../CurveTargetZone";
import { HingeGridPositionPoints } from "../HingeGrid";

describe("CurveTargetZone", () => {
  it("renders correctly", () => {
    const startPosition = HingeGridPositionPoints.centerLeft;
    const endPosition = HingeGridPositionPoints.centerRight;
    const midPointPosition = getMidPointPosition(startPosition, endPosition);
    const targetZoneAngle = 15;
    const sensorAngle = 30;
    const tolerance = 20;

    const vectorPoints = calculateControlPoints(
      startPosition,
      midPointPosition,
      targetZoneAngle,
      sensorAngle,
      tolerance
    );
    const { container } = render(
      <CurveTargetZone
        isInTargetZone={false}
        startHingePosition={startPosition}
        endHingePosition={endPosition}
        controlPointOuter={vectorPoints.controlPointOuter}
        controlPointInner={vectorPoints.controlPointInner}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
