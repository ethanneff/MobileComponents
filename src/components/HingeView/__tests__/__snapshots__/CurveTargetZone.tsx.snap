// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CurveTargetZone renders correctly 1`] = `
<View
  collapsable={true}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RNSVGPath
    d="M 550 880 Q 800 996.5769145387496 1050 880 M 1050 880 Q 800 901.872165881481 550 880 Z"
    fill={4293783794}
    fillOpacity={1}
    fillRule={1}
    matrix={
      Array [
        1,
        0,
        0,
        1,
        0,
        0,
      ]
    }
    opacity={1}
    originX={0}
    originY={0}
    propList={
      Array [
        "fill",
      ]
    }
    rotation={0}
    scaleX={1}
    scaleY={1}
    skewX={0}
    skewY={0}
    stroke={null}
    strokeDasharray={null}
    strokeDashoffset={null}
    strokeLinecap={0}
    strokeLinejoin={0}
    strokeMiterlimit={4}
    strokeOpacity={1}
    strokeWidth={1}
    vectorEffect={0}
    x={0}
    y={0}
  />
</View>
`;
